### PR TITLE
Add spec for `Enumerator#product`

### PR DESCRIPTION
**Pull request by AI-Mozi**
Hey!
It's my first contribution to this repo, so I'm not sure if i did everything right.
Also not sure if these specs make sense and are sufficient.

I'll be happy to make any adjustments as needed :D

### CODE REVIEW

File: core/enumerator/product_spec.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+require_relative '../../spec_helper'
+
+ruby_version_is "3.2" do
+  describe "Enumerator.product" do
+    it "creates an infinite enumerator" do
+      enum = Enumerator.product(1.., ["A", "B"])
+      enum.take(5).should == [[1, "A"], [1, "B"], [2, "A"], [2, "B"], [3, "A"]]
+    end
```
Comment (andrykonchin) [inline core/enumerator/product_spec.rb:8]: I would consider having the following test cases for the `Enumerator.product` method:
- the main test case IMHO should state that `Enumerator.product` "creates" a Cartesian product of enumerators. In the simplest case arguments are finite enumerators
- also we can ensure it accepts a list of enumerators of any length, so such a test could call `Enumerator.product` with 1, 2, 3, 4 etc arguments
- next, a special case with infinite enumerators (it's already present, but it's just a corner case)
- next, there might be a special case with 0 arguments
- a case with a block (it's already present)
- next, a test that checks a class of response, it's `Enumerator::Product`
- next, result object (instance of `Enumerator::Product`) has a `size` method, so we can check that it has a correct value when passed finite enumerators as arguments and it's `Float::INFINITY` when some argument is infinite enumerator
- also we can check that passed enumerators are being iterated with `#each_entry` method.
- next we can check whether arguments are implicitly type casted or the only requirement is to support the `#each_entry` method. And what happens when invalid arguments are passed

You can looks at the existing MRI specs - https://github.com/ruby/ruby/pull/6197/files#diff-a064fc14a9329854295154645b542284950529d043c261b19ca71dfc3ab7a4c9
